libfwupd/fwupd-client.c

```diff
@@ -284,22 +294,57 @@ fwupd_client_ensure_networking (FwupdClient *self, GError **error)
 				     "user agent unsuitable; fwupd version required");
 		return FALSE;
 	}
+	return TRUE;
+}
+
+static int
+fwupd_client_progress_callback_cb (void *clientp,
+				   curl_off_t dltotal,
+				   curl_off_t dlnow,
+				   curl_off_t ultotal,
+				   curl_off_t ulnow)
+{
+	FwupdClient *self = FWUPD_CLIENT (clientp);
+
+	/* calculate percentage */
+	if (dlnow <= dltotal) {
+		guint percentage = (guint) ((100 * dlnow) / dltotal);
```

Comment:
`curl_off_t` is a signed type; perhaps it should have some non-negative clamping logic before casting to `guint`? Also, should division-by-0 a concern on the `dltotal` here?

Reply:
Hmm, very valid. Also, we use the same function for uploads but don't look at ultotal at all.

Reply:
@lucab how about something like this?

    /* calculate percentage */
    if (dltotal >= 0 && dlnow >= 0 && dlnow <= dltotal) {
        guint percentage = (guint) ((100 * dlnow) / dltotal);
        g_debug ("download progress: %u%%", percentage);
        fwupd_client_set_percentage (self, percentage);
    } else if (ultotal >= 0 && ulnow >= 0 && ulnow <= ultotal) {
        guint percentage = (guint) ((100 * ulnow) / ultotal);
        g_debug ("upload progress: %u%%", percentage);
        fwupd_client_set_percentage (self, percentage);
    }

Reply:
:+1: but I think you'd want a `> 0` on both `*total` conditions (instead of the `>= 0`).

Reply:
Fixed, ta.